corehq/messaging/scheduling/scheduling_partitioned/models.py

```diff
@@ -199,19 +224,51 @@ def handle_current_event(self):
         recipient_count = 0
         for recipient in self.expand_recipients():
             recipient_count += 1
-            content.send(recipient, logged_event)
 
-        # As a precaution, always explicitly move to the next event after processing the current
-        # event to prevent ever getting stuck on the current event.
-        self.memoized_schedule.move_to_next_event(self)
-        self.memoized_schedule.move_to_next_event_not_in_the_past(self)
+            #   The framework will retry sending a non-processed schedule instance
+            # once every hour.
+
+            #   If we are processing a long list of recipients here and an error
+            # occurs half-way through, we don't want to reprocess the entire list
+            # of recipients again when the framework retries it an hour later.
+
+            #   So we use a non-blocking lock tied to the event due time and recipient
+            # to make sure that we don't try resending the same content to the same
+            # recipient more than once in the event of a retry.
+
+            #   If we succeed in sending the content, we don't release the lock so
+            # that it won't retry later. If we fail in sending the content, we release
+            # the lock so that it will retry later.
+
+            lock = self.get_content_send_lock(client, recipient)
+            if lock.acquire(blocking=False):
+                try:
+                    content.send(recipient, logged_event)
+                except:
+                    # Release the lock if an error happened so that we can try sending
+                    # to this recipient again later.
+                    lock.release()
+                    raise
```

Comment:
An exception will break out of the `for` loop. If the content for a given recipient consistently fails then it will prevent all following recipients from being processed. Would it make sense to handle the exception inside the loop so others can be processed in the same loop?

Reply:
@millerdev the approach i've taken in the reminders framework (as well as the sms framework) is that exceptions are only allowed to be raised when it's a fail-hard type of error that needs to be retried (for example when a db connection drops out).

all other errors, such as different types of configuration errors, are caught and logged in the messaging history models without raising an exception. so basically we shouldn't have a situation where one recipient's error is holding up the rest consistently, and if we do then we'd get a 500 for it and it probably means we should catch the error and log it in the messaging history models if it's the kind of thing that won't benefit from being retried (examples of that are the calls to `.error()` in the content models: https://github.com/dimagi/commcare-hq/blob/0c4bc738463d095d6b14d01e13053f0be731b64a/corehq/messaging/scheduling/models/content.py)